packages/graph/hash_graph/lib/graph/src/store/postgres/query/table.rs

```diff
@@ -90,52 +345,144 @@ impl Transpile for Table {
+            Column::EntityTypePropertyTypeReferences(column) => column.transpile(fmt),
+            Column::EntityTypeEntityTypeReferences(column) => column.transpile(fmt),
         }
     }
 }
 
-/// A column available in the database.
-#[derive(Debug, Clone, PartialEq, Eq, Hash)]
-pub struct Column<'param> {
-    pub table: Table,
-    pub access: ColumnAccess<'param>,
+#[derive(Debug, Copy, Clone, PartialEq, Eq, Hash)]
+pub enum Relation {
+    DataTypeIds,
+    PropertyTypeIds,
+    EntityTypeIds,
+    PropertyTypeDataTypeReferences,
+    PropertyTypePropertyTypeReferences,
+    EntityTypePropertyTypeReferences,
+    EntityTypeInheritance,
+    EntityType,
+    LeftEndpoint,
+    RightEndpoint,
+    OutgoingLink,
+    IncomingLink,
 }
 
-impl Transpile for Column<'_> {
-    fn transpile(&self, fmt: &mut fmt::Formatter) -> fmt::Result {
-        self.table.transpile(fmt)?;
-        fmt.write_char('.')?;
-        self.access.transpile(fmt)
+impl Relation {
+    pub const fn joins(self) -> &'static [(Column<'static>, Column<'static>)] {
+        match self {
+            Self::DataTypeIds => &[(
+                Column::DataTypes(DataTypes::VersionId),
+                Column::TypeIds(TypeIds::VersionId),
+            )],
+            Self::PropertyTypeIds => &[(
+                Column::PropertyTypes(PropertyTypes::VersionId),
+                Column::TypeIds(TypeIds::VersionId),
+            )],
+            Self::EntityTypeIds => &[(
+                Column::EntityTypes(EntityTypes::VersionId),
+                Column::TypeIds(TypeIds::VersionId),
+            )],
+            Self::PropertyTypeDataTypeReferences => &[
+                (
+                    Column::PropertyTypes(PropertyTypes::VersionId),
+                    Column::PropertyTypeDataTypeReferences(
+                        PropertyTypeDataTypeReferences::SourcePropertyTypeVersionId,
+                    ),
+                ),
+                (
+                    Column::PropertyTypeDataTypeReferences(
+                        PropertyTypeDataTypeReferences::TargetDataTypeVersionId,
+                    ),
+                    Column::DataTypes(DataTypes::VersionId),
+                ),
+            ],
+            Self::PropertyTypePropertyTypeReferences => &[
+                (
+                    Column::PropertyTypes(PropertyTypes::VersionId),
+                    Column::PropertyTypePropertyTypeReferences(
+                        PropertyTypePropertyTypeReferences::SourcePropertyTypeVersionId,
+                    ),
+                ),
+                (
+                    Column::PropertyTypePropertyTypeReferences(
+                        PropertyTypePropertyTypeReferences::TargetPropertyTypeVersionId,
+                    ),
+                    Column::PropertyTypes(PropertyTypes::VersionId),
+                ),
+            ],
+            Self::EntityTypePropertyTypeReferences => &[
```

Comment:
Question for these match arms. Is the right side of these the joins in either direction? Or how should I interpret the arrays?

Reply:
Yes, this will first join the `entity_type_property_type_references` table on `entity_types.version_id = entity_type_property_type_references.source_entity_type_version_id`, and then `property_types` table on `entity_type_property_type_references.target_property_type_version_id = entity_types.version_id`

Reply:
I see, thanks for explaining :)